### ISSUE TITLE
APIv2: prototext: unmarshal does not preserve negative 0

### ISSUE THREAD

**Issue by neild** [golang/protobuf#1011]
Text unmarshaling does not preserve the sign of floating point `-0`.

Test case (in `encoding/prototext/decode_test.go`):

```
        }, {
                desc:         "float negative zero",
                inputMessage: &pb3.Scalars{},
                inputText:    "s_float: -0",
                wantMessage: &pb3.Scalars{
                        SFloat: float32(math.Copysign(0, -1)),
                },
        }, {
```

https://oss-fuzz.com/testcase-detail/5683024310304768

**Comment by dsnet** [comment]
Quite possible this bug exists for protojson too.

**Comment by neild** [comment]
Unmarshaling does preserve negative zero when the number is unambiguously floating point: `-.0` or `-0f`.

Marshaling encodes a negative zero as a plain `-0`, which unmarshal will interpret as an integer, discarding the sign.

I'm not sure what the correct behavior is: Should unmarshal preserve the sign of `-0`? (Requires understanding that `-0` might be an integer or a float.) Or should marshal encode floating point `-0` as `-.0`/`-0.0`/`-0f`?

**Comment by neild** [comment]
`protojson` is fine, because JSON only has floating point numbers so there's no point where the sign bit is lost.

**Comment by puellanivis** [comment]
I think this is a situation where we should look at what the C++ implementation does? At least until there is a defined standard to follow.

If C++ is printing just a `-0` then we have to consider we might be getting simple `-0` as an input, no matter what we output from Marshal. So we might have to find some way to keep the value ambiguously, until it is assigned. :weary: We might be able to short-circuit something and dictate that `-0` is always a floating point value? Then only when it is assigned into an integer field, it gets properly cast into the expected non-negative 0.

On the other side of the issue, I would probably prefer `-0.0` as output from Marshal. As it is the most abundantly clear as to what it is and why it is.

**Comment by neild** [comment]
Agreed that we should look to C++'s behavior for guidance.

```shell
$ echo 'optional_float:-0' | protoc -Iinternal/testprotos internal/testprotos/test/test.proto --encode=goproto.proto.test.TestAllTypes | protoc -Iinternal/testprotos internal/testprotos/test/test.proto --decode=goproto.proto.test.TestAllTypes
optional_float: -0
```

I'm not certain if `protoc`'s text (de)serialization uses the standard C++ implementation or not, but it both understands and outputs `-0` for negative zero.

So I guess the minimal fix is going to be to decoding.

**Comment by cybrcodr** [comment]
I just checked C++ behavior.

The TextFormat package defined in google/protobuf/text_format.h does print -0 and parsing does preserve the sign bit.

For the JSON serialization lib in google/protobuf/util/json_util.h, it does print -0 as well.  Parsing double does preserve the sign bit, but parsing float somehow didn't.

https://golang.org/cl/204602  which rewrites internal/encoding/text does fix this for prototext already.  I've incorporated the tests from neild@'s https://golang.org/cl/217501.

And current protojson does not have this issue.
